nested-json-parser/src/main/java/org/apache/druid/data/input/nested/NestedJSONInputRowParser.java

```diff
@@ -84,7 +84,7 @@ public NestedJSONInputRowParser(
     // Maybe assert that pivotSpec's dimension names exist in the
     // existing specified dimensions?
     Preconditions.checkArgument(
-        parseSpec.getDimensionsSpec().hasCustomDimensions(),
+        parseSpec.getDimensionsSpec().hasFixedDimensions(),
```

Comment:
```
public boolean hasCustomDimensions()
{
   return !(dimensions == null || dimensions.isEmpty());
}
```
changed to:
```
public boolean hasFixedDimensions()
{
  return dimensions != null && !dimensions.isEmpty() && !useSchemaDiscovery && !includeAllDimensions;
}
```
"Rename DimensionsSpec#hasCustomDimensions to hasFixedDimensions and change the meaning subtly: it now
returns true if the DimensionsSpec represents an unchanging list of dimensions, or false if there is
some discovery happening. This is what call sites had expected anyway."